File: php/src/CodingInterview/Application/Usecase/Asset/GetAssetUsecase.php
<?php

declare(strict_types=1);

namespace Folio\CodingInterview\Application\Usecase\Asset;

use Folio\CodingInterview\Application\Repository\AccountRepository;
use Folio\CodingInterview\Domain\BigDecimal;
use Folio\CodingInterview\Domain\StockSymbol;
use Folio\CodingInterview\Domain\UserId;

final class GetAssetUsecaseInput
{
    public function __construct(public readonly UserId $userId) {}
}

final class GetAssetStockOutput
{
    public function __construct(
        public readonly StockSymbol $symbol,
        public readonly BigDecimal $amountJpy,
    ) {}
}

final class GetAssetUsecaseOutput
{
    /** @param GetAssetStockOutput[] $stocks */
    public function __construct(
        public readonly BigDecimal $cashAmount,
        public readonly array $stocks,
    ) {}
}

final class GetAssetUsecase
{
    public function __construct(
        private readonly AccountRepository $accountRepository,
    ) {}

    public function run(GetAssetUsecaseInput $input): GetAssetUsecaseOutput
    {
        $account = $this->accountRepository->find($input->userId);
        if ($account === null) {
            throw new GetAssetUsecaseUserNotFoundException();
        }
        $stocks = [];
        foreach ($account->stocks as $e) {
            $stocks[] = new GetAssetStockOutput($e->symbol, $e->amountJpy);
        }
        return new GetAssetUsecaseOutput($account->cash, $stocks);
    }
}

File: scala/src/main/scala/folio/codinginterview/application/usecase/asset/GetAssetUsecase.scala
package folio.codinginterview.application.usecase.asset

import folio.codinginterview.application.repository.AccountRepository
import folio.codinginterview.domain.StockSymbol
import folio.codinginterview.domain.UserId
import scala.concurrent.ExecutionContext
import scala.concurrent.Future

final case class GetAssetUsecaseInput(userId: UserId)

final case class GetAssetStockOutput(symbol: StockSymbol, amountJpy: BigDecimal)

final case class GetAssetUsecaseOutput(
    cashAmount: BigDecimal,
    stocks: Seq[GetAssetStockOutput]
)

sealed trait GetAssetUsecaseException extends RuntimeException
object GetAssetUsecaseException {
  case object UserNotFound extends GetAssetUsecaseException
}

final class GetAssetUsecase(
    accountRepository: AccountRepository
)(using ec: ExecutionContext) {
  def run(input: GetAssetUsecaseInput): Future[GetAssetUsecaseOutput] = {
    for {
      maybeAccount <- accountRepository.find(input.userId)
      account <- maybeAccount match {
        case Some(a) => Future.successful(a)
        case None    => Future.failed(GetAssetUsecaseException.UserNotFound)
      }
    } yield {
      val stocks = account.stocks.map { e =>
        GetAssetStockOutput(e.symbol, e.amountJpy)
      }
      GetAssetUsecaseOutput(account.cash, stocks)
    }
  }
}
